### ISSUE TITLE
Transformation of client request before use

### ISSUE THREAD

**Issue by jonathanstowe** [croservices/cro-http#153]
In order to make, for example, a [Signed AWS request](https://docs.aws.amazon.com/general/latest/gr/sigv4_signing.html) it would be handy if the Cro::HTTP::Client could have a mechanism that a Request -> Request transform ( or transforms) can be applied after the Request has been fully assembled and before the actual http request is made, such that new headers can be computed and added based on the headers and body of the request.

On the face of it it looks like it could be inserted in `.request` immediately before the `$pipeline.send-request` is called as that will be the last place the headers could be changed.

**Comment by jonathanstowe** [comment]
Ah no, looking more deeply in the code it would seem the best place to add this would be in `build-pipeline` *after* the body serializers and before the request serializer is pushed.  I'll have a poke.

**Comment by jnthn** [comment]
An idea I had a long time back - but we never yet implemented - was to have middleware supported on the HTTP client, with an API like that on the server side. That would include being able to circumvent the entire request on the network and provide a response directly, so we could use it to implement all manner of things (local caching, circuit breaker, avoiding requests that would be beyond a rate limit, etc.) Then an API at construction time with `before`/`after`, again matching the HTTP server approach. I think the `Cro::HTTP::Middleware` interface should fit with client side as well as server side - at least, that was my hope.

**Comment by jonathanstowe** [comment]
Yeah on the face of it that wouldn't require too much new code either as most of the parts from Server seem re-usable at first glance.
